src/blas/KokkosBlasLapack_trtri.hpp

```diff
@@ -0,0 +1,144 @@
+///                   On entry, A
+///                   On successful exit, inv(A)
+/// \return           0 upon success, 
+//                    i if the i-th diagonal elemet of A is zero, A is singular,
+//                    and the inversion could not be completed.
+// source: https://software.intel.com/en-us/mkl-developer-reference-c-trtri
+template<class AViewType>
+int
+trtri (const char uplo[],
+      const char diag[],
+      const AViewType& A)
+{
+
+  static_assert (Kokkos::Impl::is_view<AViewType>::value,
+                 "AViewType must be a Kokkos::View.");
+  static_assert (static_cast<int> (AViewType::rank) == 2,
+                 "AViewType must have rank 2.");
+
+  // Check validity of indicator argument
+  bool valid_uplo  = (uplo[0] == 'U' ) || (uplo[0] == 'u' )||
+                     (uplo[0] == 'L' ) || (uplo[0] == 'l' );
+  bool valid_diag  = (diag[0] == 'U' ) || (diag[0] == 'u' )||
+                     (diag[0] == 'N' ) || (diag[0] == 'n' );
+
+  if(!valid_uplo) {
+    std::ostringstream os;
+    os << "KokkosBlasLapack::trtri: uplo = '" << uplo[0] << "'. " <<
+      "Valid values include 'U' or 'u' (A is upper triangular), "
+      "'L' or 'l' (A is lower triangular).";
+    Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+  if(!valid_diag) {
+    std::ostringstream os;
+    os << "KokkosBlasLapack::trtri: diag = '" << diag[0] << "'. " <<
+      "Valid values include 'U' or 'u' (the diagonal of A is assumed to be unit), "
+     "'N' or 'n' (the diagonal of A is assumed to be non-unit).";
+    Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+
+  int64_t A_m = A.extent(0);
+  int64_t A_n = A.extent(1);
+
+  // Return if degenerated matrices are provided
+  if((A_m == 0) || (A_n == 0))
+    return 0; // This is success as the inverse of a matrix with no elements is itself.
+
+  // Ensure that the dimensions of A match and that we can legally perform A*B
+  // or B*A
+  if (A_m != A_n) {
+    std::ostringstream os;
+    os << "KokkosBlasLapack::trtri: Dimensions of A do not match,"
+       << " A: " << A.extent(0) << " x " << A.extent(1);
+    Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+
+  // Create A matrix view type alias
+  using AViewInternalType = Kokkos::View<typename AViewType::const_value_type**,
+                           typename AViewType::array_layout,
+                           typename AViewType::device_type,
+                           Kokkos::MemoryTraits<Kokkos::Unmanaged> >;
+
+  // This is the return value type and should always reside on host
+  using RViewInternalType = Kokkos::View<int,
+                           typename AViewType::array_layout,
+                           Kokkos::HostSpace,
+                           Kokkos::MemoryTraits<Kokkos::Unmanaged> >;
+
+  int result;
+  RViewInternalType R = RViewInternalType(&result);
+
+  KokkosBlas::Impl::TRTRI<RViewInternalType, AViewInternalType>::trtri (R, uplo, diag, A);
+  Kokkos::fence();
```

Comment:
I get why you need a fence here, but it seems like users either don't care about the error code, or they want it left on device to avoid an unnecessary fence or (synchronous) copy back to host.  Your interface here forces that result back to host, even if users plan on ignoring the error code.